doc/source/handling-dependencies.rst

```diff
@@ -0,0 +1,256 @@
+* Your Ray script may import some files outside of the script (i.e., via relative imports)
+
+
+One frequent problem when running on a cluster is that Ray expects these "dependencies" to exist on each Ray node. Otherwise, you may run into different issues such as <TODO ADD ISSUE LINK>.
+
+
+To address this problem, you can leverage Ray's **Runtime environments**.
+
+
+Concepts
+--------
+
+- **Ray Application**
+
+- **Local machine** and **Cluster**.  The recommended way to connect to a remote Ray cluster is to use :ref:`Ray Client<ray-client>`, and we will call the machine running Ray Client your *local machine*.  Note: you can also start a single-node Ray cluster on your local machine---in this case your Ray cluster is not really “remote”, but any comments in this documentation referring to a “remote cluster” will also apply to this setup.
+
+- **Files**: These are the files that your Ray application needs to run.  These can include code files or data files.  For a development workflow, these might live on your local machine, but when it comes time to run things at scale, you will need to get them to your remote cluster.  For how to do this, see :ref:`Workflow: Local Files<workflow-local-files>` below.
+
+- **Packages**: These are external libraries or executables required by your Ray application, often installed via ``pip`` or ``conda``.
+
+.. Alternatively, you can prepare your Ray cluster's environment when your cluster nodes start up, and modify it later from the command line.
+.. Packages can be installed using ``setup_commands`` in the Ray Cluster configuration file (:ref:`docs<cluster-configuration-setup-commands>`) and files can be pushed to the cluster using ``ray rsync_up`` (:ref:`docs<ray-rsync>`).
+
+
+.. _runtime-environments:
+
+Runtime Environments
+--------------------
+
+.. note::
+
+    This feature requires a full installation of Ray using ``pip install "ray[default]>=1.4"``, and is currently only supported on macOS and Linux.
+
+A **runtime environment** describes the dependencies your Ray application needs to run, including :ref:`files, packages, environment variables, and more <runtime-environments-api-ref>`.
+
+Runtime environment enable the transition of running of your Ray application on your local machine to a remote cluster, without any code changes or manual setup.
+
+..
+  TODO(architkulkarni): run working_dir doc example in CI
+
+.. code-block:: python
+
+    runtime_env = {"working_dir": "/data/my_files", "pip": ["requests", "pendulum==2.1.2"]}
+
+    # RLIAW: ADD example of this actually working
+
+.. literalinclude:: ../examples/doc_code/runtime_env_example.py
+   :language: python
+   :start-after: __runtime_env_conda_def_start__
+   :end-before: __runtime_env_conda_def_end__
+
+Jump to the :ref:`API Reference<runtime-environments-api-ref>`.
+
+
+There are two primary ways of using runtime environments:
+
+* Per Ray Job (TODO: define job) <link down>
+* Per Ray Task/Actor, within a job <link down>
```

Comment:
TODOs here?

Reply:
Sorry, this is in kind of an intermediate state after going over it with Richard.  I still need to make more changes and clean things up